src/materialized.rs

```diff
@@ -95,15 +98,31 @@ pub trait Materialized: ListingTableLike {
     fn query(&self) -> LogicalPlan;
 }
 
+/// Register a [`Materialized`] implementation in this registry.
+/// This allows `cast_to_materialized` to easily downcast a [`TableProvider`]
+/// into a `Materialized` where possible.
+///
+/// Note that this will also register `T` as a [`ListingTableLike`].
+pub fn register_materialized<T: Materialized>() {
+    TABLE_TYPE_REGISTRY.register_materialized::<T>();
+}
+
+/// Attempt to cast the given TableProvider into a [`Materialized`].
+/// If the table's type has not been registered using [`register_materialized`], will return `None`.
+pub fn cast_to_materialized(table: &dyn TableProvider) -> Option<&dyn Materialized> {
+    TABLE_TYPE_REGISTRY.cast_to_materialized(table)
+}
+
 type Downcaster<T> = Arc<dyn Fn(&dyn Any) -> Option<&T> + Send + Sync>;
 
-/// A registry for implementations of [`ListingTableLike`], used for downcasting
-/// arbitrary TableProviders into `dyn ListingTableLike` where possible.
+/// A registry for implementations of library-defined traits, used for downcasting
+/// arbitrary TableProviders into `ListingTableLike` and `Materialized` trait objects where possible.
 ///
-/// This is used throughout the crate as a singleton to store all known implementations of `ListingTableLike`.
-/// By default, [`ListingTable`] is registered.
+/// This is used throughout the crate as a singleton to store all known implementations of `ListingTableLike` and `Materialized`.
+/// By default, [`ListingTable`] is registered as a `ListingTableLike`.
 struct TableTypeRegistry {
     listing_table_accessors: DashMap<TypeId, (&'static str, Downcaster<dyn ListingTableLike>)>,
+    materialized_accessors: DashMap<TypeId, (&'static str, Downcaster<dyn Materialized>)>,
```

Comment:
this is an interesting approach -- very cool. I wonder if there is anyhing we can do to the DataFusion APIs to make this easier to use

Reply:
Thanks, yeah it's a bit unfortunate, but it's the only way I could think of to support "trait downcasting" in Rust code. In Go I know it is assisted by some runtime type information handled by the language itself. The only case I know of where someone has implemented their own such mechanism is the [`bevy_reflect`](https://docs.rs/bevy_reflect/latest/bevy_reflect/index.html) crate, which seems to support trait downcasting here: https://docs.rs/bevy_reflect/latest/bevy_reflect/attr.reflect_trait.html